pytorch_lightning/callbacks/timer.py

```diff
@@ -0,0 +1,117 @@
+    if the given time limit is reached.
+
+    Args:
+        duration: A string in the format HH:MM:SS (hours, minutes seconds), or a :class:`datetime.timedelta`.
+        interval: Determines if the interruption happens on epoch level or mid-epoch.
+            Can be either `epoch` or `step`.
+        verbose: Set this to ``False`` to suppress logging messages.
+
+    Raises:
+        MisconfigurationException:
+            If ``interval`` is not one of the supported choices.
+
+    Example::
+        from pytorch_lightning import Trainer
+        from pytorch_lightning.callbacks import Timer
+
+        # stop training after 12 hours
+        timer = Timer(duration="12:00:00")
+
+        # or provide a datetime.timedelta
+        from datetime import timedelta
+        timer = Timer(duration=timedelta(weeks=1))
+
+        # force training to stop after given time limit
+        trainer = Trainer(callbacks=[timer])
+    """
+
+    INTERVAL_CHOICES = ("epoch", "step")
+
+    def __init__(self, duration: Union[str, timedelta], interval: str = "step", verbose: bool = True):
+        super().__init__()
+        if isinstance(duration, str):
+            hms = datetime.strptime(duration.strip(), "%H:%M:%S")
+            duration = timedelta(hours=hms.hour, minutes=hms.minute, seconds=hms.second)
+        if interval not in self.INTERVAL_CHOICES:
+            raise MisconfigurationException(
+                f"Unsupported parameter value `Timer(interval={interval})`. Possible choices are:"
+                f" {', '.join(self.INTERVAL_CHOICES)}"
+            )
+        self._duration = duration
+        self._interval = interval
+        self._verbose = verbose
+        self._start_time = None
+        self._offset = timedelta()
+
+    @property
+    def start_time(self):
+        return self._start_time
+
+    @property
+    def time_elapsed(self) -> timedelta:
+        if self._start_time is None:
+            return self._offset
+        return datetime.now() - self._start_time + self._offset
+
+    @property
+    def time_remaining(self) -> timedelta:
+        return self._duration - self.time_elapsed
+
+    def on_train_start(self, trainer, *args, **kwargs) -> None:
+        self._start_time = datetime.now()
+
+    def on_train_batch_end(self, trainer, *args, **kwargs) -> None:
+        if self._interval != "step":
+            return
+        self._check_time_remaining(trainer)
+
+    def on_train_epoch_end(self, trainer, *args, **kwargs) -> None:
+        if self._interval != "epoch":
+            return
+        self._check_time_remaining(trainer)
+
+    def on_save_checkpoint(self, trainer, pl_module, checkpoint: Dict[str, Any]) -> Dict[str, Any]:
+        return {
+            "time_elapsed": self.time_elapsed,
+        }
+
+    def on_load_checkpoint(self, callback_state: Dict[str, Any]):
+        self._offset = callback_state.get("time_elapsed", timedelta())
```

Comment:
why the different names offset and time_elapsed? choose one?

Reply:
I choose the name offset because it's used differently than just the time elapsed.
I'm not saving the time elapsed directly as an attribute because it is continuously changing, while the offset is needed to be able to resume the time when resuming training from a checkpoint.